Tree: update ReadMes (#12141)
microsoft/FluidFramework: packages/dds/tree/README.md

    end
    package-->runtime["Fluid runtime"]
```

> # Open Design Questions

Reviewer: I think this might deserve its own page so the README can be a more atomic view of the current state of the tree.

Author: This readme tries to detail the actual goals of this package, and how its architected to meet those goals. The known design issues where its design (as covered in this document) doesn't work seem in scope.

Also I think the fact that significant parts of the codebase don't really work due to design issue is part of the current state of this package. Note that the design issues here all impact the architectural role of top level submodules in this package in a way that when fixes will likely require changes to the architectural details covered above.

I'm updating the description to clarify this.

```suggestion
# Open Design Questions

The design issues here all impact the architectural role of top-level modules in this package in a way that when fixed will likely require changes to the architectural details covered above.
Smaller scoped issues which will not impact the overall architecture should be documented in more localized locations.
```

## How should indexes relate to branches?
Some possible options:

>  - Use copy on write in indexes, and keep all needed indexes for all needed revision within edit-manager. Provide all relevant indexes to change rebaser. Maybe allow change rebaser to compute intermediate indexes as needed.

Reviewer: ```suggestion
 - Use copy on write in indexes, and keep all needed indexes for all needed revisions within edit-manager. Provide all relevant indexes to change rebaser. Maybe allow change rebaser to compute intermediate indexes as needed.
```

Also nit: edit-manager having a hyphen but not change rebaser is a bit weird to me. I'm also wondering if they should both be referred to by class name e.g. `ChangeRebaser`.

Author: This is referring to edit-manager, the folder/module. It could instead refer to EditManager, the class, but this file (the top level readme) talks about these modules by their module name in the diagrams above, so I figured it would be more clear this way. ChangeRebaser should be fixed though.
```suggestion
 - Use copy on write in indexes, and keep all needed indexes for all needed revisions within edit-manager. Provide all relevant indexes to `ChangeRebaser`. Maybe allow `ChangeRebaser` to compute intermediate indexes as needed.
```

 - Keep a single index, and adjust it to the needed location in the branch tree as needed using deltas.
 - Keep multiple indexes, one at each branch head, updated via mutation.
 - Keep a single reference index (maybe after the latest sequenced edit), and make delta indexes referencing it for the other required branches.
 - Something else?

## How should specialized sub-tree handling compose?

Applications should have a domain model that can mix editable tree nodes with custom implementations as needed.
Custom implementations should probably be able to be projections of editable trees, the forest content (via cursors), and updated via either regeneration from the input, or updated by a delta.

> This is important for performance / scalability, and might be how we do virtualization (maybe subtrees that arn't downloaded are just one custom representation?).

Reviewer: ```suggestion
This is important for performance/scalability and might be how we do virtualization (maybe subtrees that aren't downloaded are just one custom representation?).
```

This might also be the layer at which we hook up schematize.

> Alternatively it might be an explicitly two phase setup (schematize then normalize), but we might share logic between the two and have non-copying bypasses.

Reviewer: ```suggestion
Alternatively, it might be an explicitly two-phase setup (schematize then normalize), but we might share logic between the two and have non-copying bypasses.
```
